client/src/main/java/org/eclipse/hono/client/impl/CommandTargetMapperImpl.java

```diff
@@ -0,0 +1,160 @@
+                            .compose(resultJson -> {
+                                final JsonArray instancesArray = resultJson.getJsonArray(DeviceConnectionConstants.FIELD_ADAPTER_INSTANCES);
+                                if (instancesArray == null || instancesArray.isEmpty()) {
+                                    LOG.error("Invalid result JSON; field '{}' is null or empty", DeviceConnectionConstants.FIELD_ADAPTER_INSTANCES);
+                                    return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+                                }
+
+                                final JsonObject targetInstanceObject;
+                                try {
+                                    if (instancesArray.size() == 1) {
+                                        targetInstanceObject = instancesArray.getJsonObject(0);
+                                    } else {
+                                        targetInstanceObject = chooseTargetGatewayAndAdapterInstance(instancesArray);
+                                    }
+                                } catch (final ClassCastException e) {
+                                    LOG.error("Invalid result JSON", e);
+                                    return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+                                }
+                                final String targetDevice = targetInstanceObject.getString(DeviceConnectionConstants.FIELD_PAYLOAD_DEVICE_ID);
+                                final String targetAdapterInstance = targetInstanceObject.getString(DeviceConnectionConstants.FIELD_ADAPTER_INSTANCE_ID);
+                                if (targetDevice == null || targetAdapterInstance == null) {
+                                    LOG.error("Invalid result JSON, missing target device and/or adapter instance");
+                                    return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+                                }
+                                if (!targetDevice.equals(deviceId)) {
+                                    // target device is a gateway
+                                    if (!viaGateways.contains(targetDevice)) {
+                                        LOG.error("Invalid result JSON, target gateway {} is not in via gateways list", targetDevice);
+                                        return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+                                    }
+                                    span.setTag(MessageHelper.APP_PROPERTY_GATEWAY_ID, targetDevice);
+                                }
+                                span.finish();
+                                return Future.succeededFuture(targetInstanceObject);
+                            });
```

Comment:
Good point.

> we could also move it up one level in the compose hierarchy FMPOV ...

That won't work because of the `viaGateways` list initialized in one compose block.

Reply:
I've put it in a separate method and improved logging/tracing.